assets/js/util/storage.js

```diff
@@ -0,0 +1,142 @@
+/**
+ * Storage abstraction.
+ *
+ * Site Kit by Google, Copyright 2019 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Detects whether browser storage is both supported and available.
+ *
+ * @param {string} type Browser storage to test. ex localStorage or sessionStorage.
+ * @return {boolean} True if the given storage is available, false otherwise.
+ */
+export const storageAvailable = ( type ) => {
+	const storage = window[ type ];
+	if ( ! storage ) {
+		return false;
+	}
+	try {
+		const x = '__storage_test__';
+
+		storage.setItem( x, x );
+		storage.removeItem( x );
+		return true;
+	} catch ( e ) {
+		return e instanceof DOMException && (
+
+			// everything except Firefox
+			22 === e.code ||
+
+			// Firefox
+			1014 === e.code ||
+
+			// test name field too, because code might not be present
+			// everything except Firefox
+			'QuotaExceededError' === e.name ||
+
+			// Firefox
+			'NS_ERROR_DOM_QUOTA_REACHED' === e.name ) &&
+
+			// acknowledge QuotaExceededError only if there's something already stored
+			0 !== storage.length;
+	}
+};
+
+/**
+ * Detects whether and which persistent storage is available.
+ *
+ * @return {mixed} Either 'sessionStorage', 'localStorage', or undefined.
+ */
+const getStorageName = () => {
+	if ( storageAvailable( 'sessionStorage' ) ) {
+		return 'sessionStorage';
+	}
+
+	if ( storageAvailable( 'localStorage' ) ) {
+		return 'localStorage';
+	}
+
+	return undefined;
+};
+
+export const storage = {
```

Comment:
1. This should get a `clear` method. This should then be used in (or in place of) `clearAppLocalStorage`.
1. I think we should refrain from using methods that are not part of the [`Storage`](https://developer.mozilla.org/en-US/docs/Web/API/Storage#Methods) interface. Unless we're polyfilling the behavior of the methods themselves for some reason then `storage` can simply be set to the real `window[implementation]`. Then, only the no-op Storage implementation needs to be written, which should be super simple. I think this would simplify the implementation here a good deal. `isAvailable` should really not be necessary as this would fall to the null/no-op case 😄 .
Pseudo code:
```js
export const storage = sessionStorage || localStorage || nullStorage;
```

There are also a few references left in the code that do not use the abstraction:
* https://github.com/google/site-kit-wp/blob/b225df85d49c114e161591c91bbdb6a3c8015995/assets/js/googlesitekit-admin.js#L30
* https://github.com/google/site-kit-wp/blob/06ed3f4ce47b0bd484282fa5048830add9717441/assets/js/components/notifications/util.js#L72

Reply:
@aaemnnosttv This makes sense. The only reason _for_ an `isAvailable()` is that we could check whether to perform related logic or not - if we can't cache the value, no need to do logic to prepare the value (although that's probably cheap enough to not care). So I'm in agreement with you for the cleanliness of the API.

The point on the few references still using `localStorage` directly I'd like to skip for now. These are for a very special case of using as little custom JS as possible since they tie in with the general admin bar script that is loaded almost everywhere to display the notification count. In that area (and all other code related with that specific cache), `localStorage` is currently hardcoded, and I'd prefer if we revisited this separately.

Reply:
> if we can't cache the value, no need to do logic to prepare the value (although that's probably cheap enough to not care)

Agreed. Unless it's performing extra http calls, or something truly expensive, it's probably not worth worrying about 👍 

> The point on the few references still using localStorage directly I'd like to skip for now. These are for a very special case of using as little custom JS as possible since they tie in with the general admin bar script that is loaded almost everywhere to display the notification count.

In these cases, can we then annotate the special case with a comment for visibility?

Reply:
@aaemnnosttv On second thought, I'd prefer to still clear the storages directly in `clearAppLocalStorage()` (that name is weird btw, but something for another day). Reason is that we should clear both caches, because there is at least a theoretical chance that different storage objects are used by the same site (e.g. if the other's capacity was at some point exhausted).